File: src/main/java/com/nhom2/MaxxSports/service/OrderService.java
package com.nhom2.MaxxSports.service;

import com.nhom2.MaxxSports.dto.request.*;
import com.nhom2.MaxxSports.dto.response.OrderResponse;
import com.nhom2.MaxxSports.entity.*;
import com.nhom2.MaxxSports.enums.*;
import com.nhom2.MaxxSports.mapper.OrderMapper;
import com.nhom2.MaxxSports.repository.*;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class OrderService {

    private final UserRepository userRepository;

    private final ProductDetailRepository
            productDetailRepository;

    private final OrderRepository
            orderRepository;

    private final PaymentRepository
            paymentRepository;

    private final OrderMapper orderMapper;

    @Transactional
    public OrderResponse checkout(
            String email,
            CheckoutRequest request
    ) {

        log.info("Checkout");

        User user = userRepository.findByEmail(email)
                .orElseThrow(() ->
                        new RuntimeException(
                                "Bạn cần phải đăng nhập"
                        ));

        Order order = Order.builder()
                .orderDate(LocalDateTime.now())
                .detailAddress(
                        request.getDetailAddress()
                )
                .user(user)
                .orderStatus(
                        request.getPaymentMethod()
                                == PaymentMethod.VNPAY
                                ? OrderStatus.CHO_THANH_TOAN
                                : OrderStatus.PENDING
                )
                .shippingMethod(
                        request.getShippingMethod()
                )
                .build();

        List<OrderDetail> orderDetails =
                new ArrayList<>();

        int totalQuantity = 0;

        double totalPrice = 0;

        for (
                CheckoutItemRequest item
                : request.getItems()
        ) {

            ProductDetail productDetail =
                    productDetailRepository
                            .findById(
                                    item.getProductDetailId()
                            )
                            .orElseThrow(() ->
                                    new RuntimeException(
                                            "Sản phẩm không tồn tại"
                                    ));

            // kiểm tra tồn kho
            if (
                    productDetail.getSoLuong()
                            < item.getQuantity()
            ) {

                throw new RuntimeException(
                        "Sản phẩm không đủ số lượng"
                );
            }

            double price =
                    productDetail.getProduct()
                            .getGia();

            double itemTotal =
                    price * item.getQuantity();

            OrderDetail orderDetail =
                    OrderDetail.builder()
                            .order(order)
                            .productDetail(productDetail)
                            .quantity(
                                    item.getQuantity()
                            )
                            .price(price)
                            .totalPrice(itemTotal)
                            .build();

            orderDetails.add(orderDetail);

            totalQuantity +=
                    item.getQuantity();

            totalPrice += itemTotal;

            // trừ số lượng tồn
            productDetail.setSoLuong(
                    productDetail.getSoLuong()
                            - item.getQuantity()
            );

            productDetailRepository
                    .save(productDetail);
        }

        order.setOrderDetails(orderDetails);

        order.setTotalQuantity(
                totalQuantity
        );

        order.setTotalPrice(
                totalPrice
        );

        orderRepository.save(order);

        Payment payment = Payment.builder()
                .order(order)
                .amount(totalPrice)
                .method(
                        request.getPaymentMethod()
                )
                .status(
                        PaymentStatus.PENDING
                )
                .transactionNo(
                        UUID.randomUUID()
                                .toString()
                )
                .build();

        paymentRepository.save(payment);

        order.setPayment(payment);

        return orderMapper.toResponse(order);
    }

    public List<OrderResponse> getMyOrders(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy user"));

        return orderRepository.findByUser(user)
                .stream()
                .map(orderMapper::toResponse)
                .toList();
    }

    public List<OrderResponse> getAllOrders() {
        return orderRepository.findAll().stream().map(orderMapper::toResponse).toList();
    }

    public String updateOrder(UpdateOrderRequest request) {
        var order= orderRepository.findById(request.getId()).orElseThrow(()->new RuntimeException("Đơn hàng không tồn tại"));

        order.setOrderStatus(request.getStatus());

        orderRepository.save(order);
        return "Cập nhập đơn hàng thành công";
    }
}

File: web-app/src/services/orderService.js
import api from '../api/api';

export const orderService = {

    // =====================================================================
    // LUỒNG USER/CLIENT (Khách mua hàng thực hiện Đặt đơn & Xem lịch sử)
    // =====================================================================

    /**
     * 1. Tiến hành đặt hàng (Checkout) và tạo hóa đơn mới
     * @param {Object} checkoutRequest - Chứa cấu trúc:
     * {
     * detailAddress: "Số nhà, tên đường, Hà Nội...",
     * shippingMethod: "FAST" hoặc "STANDARD",
     * paymentMethod: "COD" hoặc "VNPAY",
     * items: [ { productDetailId: "uuid-chuỗi", quantity: 2 }, ... ]
     * }
     * @returns {Promise<Object>} Trả về ApiResponse chứa dữ liệu OrderResponse ở mục .result
     */
    checkout: async (checkoutRequest) => {
        // Vì Backend trả về dạng ApiResponse<OrderResponse>
        const response = await api.post('/order/checkout', checkoutRequest);
        return response.data;
    },

    /**
     * 2. Khách hàng xem lịch sử toàn bộ đơn hàng cá nhân đã đặt tại MaxxSports
     * @returns {Promise<Object>} Trả về ApiResponse chứa danh sách List<OrderResponse> ở mục .result
     */
    getMyOrders: async () => {
        const response = await api.get('/order/myOrders');
        return response.data;
    },

    // =====================================================================
    // LUỒNG ADMIN (Quản trị viên quản lý, duyệt đơn và thay đổi trạng thái)
    // =====================================================================

    /**
     * 3. [Admin] Lấy toàn bộ danh sách đơn đặt hàng của toàn bộ hệ thống MaxxSports
     * @returns {Promise<Object>} Trả về ApiResponse chứa List<OrderResponse> ở mục .result
     */
    getAllOrdersForAdmin: async () => {
        const response = await api.get('/order/getAllOrders');
        return response.data;
    },

    /**
     * 4. [Admin] Cập nhật trạng thái xử lý của đơn hàng (Ví dụ: Chuyển từ PENDING -> SHIPPING -> DELIVERED)
     * @param {Object} updateOrderRequest - Chứa cấu trúc { id: "mã-đơn-hàng", status: "SHIPPING" }
     * @returns {Promise<Object>} Trả về ApiResponse chứa chuỗi "Cập nhập đơn hàng thành công" ở mục .result
     */
    updateOrderStatusByAdmin: async (updateOrderRequest) => {
        const response = await api.put('/order/status', updateOrderRequest);
        return response.data;
    }
};